src/manager/evm.rs

```diff
@@ -0,0 +1,379 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+use std::collections::HashMap;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use cid::Cid;
+use ethers::prelude::k256::ecdsa::SigningKey;
+use ethers::prelude::{abigen, Signer, SignerMiddleware};
+use ethers::providers::{Authorization, Http, Middleware, Provider};
+use ethers::signers::{LocalWallet, Wallet};
+use fvm_shared::address::Payload;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::{address::Address, econ::TokenAmount};
+use ipc_gateway::BottomUpCheckpoint;
+use ipc_sdk::subnet_id::SubnetID;
+use ipc_subnet_actor::{ConstructParams, JoinParams};
+use num_traits::ToPrimitive;
+use primitives::EthAddress;
+
+use super::subnet::SubnetManager;
+use crate::config::subnet::SubnetConfig;
+use crate::config::Subnet;
+use crate::lotus::message::ipc::SubnetInfo;
+use crate::lotus::message::wallet::WalletKeyType;
+pub type MiddlewareImpl = SignerMiddleware<Provider<Http>, Wallet<SigningKey>>;
+
+/// The majority vote percentage for checkpoint submission when creating a subnet.
+const SUBNET_MAJORITY_PERCENTAGE: u8 = 60;
+const TRANSACTION_RECEIPT_RETRIES: usize = 10;
+
+// Create type bindings for the IPC Solidity contracts
+abigen!(Gateway, "contracts/Gateway.json");
+abigen!(SubnetContract, "contracts/SubnetActor.json");
+abigen!(SubnetRegistry, "contracts/SubnetRegistry.json");
+
+pub struct EthSubnetManager<M: Middleware> {
+    eth_client: Arc<M>,
+    gateway_contract: Gateway<Arc<M>>,
+    registry_contract: SubnetRegistry<Arc<M>>,
+}
+
+#[async_trait]
+impl<M: Middleware + Send + Sync + 'static> SubnetManager for EthSubnetManager<M> {
+    async fn create_subnet(&self, _from: Address, params: ConstructParams) -> Result<Address> {
+        self.ensure_same_gateway(&params.ipc_gateway_addr)?;
+
+        let min_validator_stake = params
+            .min_validator_stake
+            .atto()
+            .to_u128()
+            .ok_or_else(|| anyhow!("invalid min validator stake"))?;
+
+        log::debug!("in create subnet");
+
+        let mut route = agent_subnet_to_evm_addresses(&params.parent)?;
+        let mut root = vec![ethers::types::Address::from_str(
```

Comment:
Why the first element of the route needs to be the zero address?

Reply:
Good one, I think I saw it before in the Solidity contracts, but now I think it's removed.

Reply:
updated